Size operator possibly returning real/double values?

When using the spicy-analyzers http analyzer, I'm seeing the following error message coming from Zeek which leads me to believe that the size operator (`|x|`) isn't returning an unsigned integer like I would expect it to.

```
analyzer error in /spicy-analyzers-0.2.22/analyzer/protocol/http/http.evt, line 20: Event parameter mismatch, cannot convert Spicy value of type 'double' to Zeek value of type 'count'
```

On the Spicy runtime side the size operator for list returns a `uint64_t` which we do not handle when mapping types in zeek/spicy-plugin. We could fix this in the plugin, but the result type already causes problems on the Spicy side where we do not handle it well, either, e.g.,

```
$ cat f.hlt
module foo {
import hilti;

global auto xs = [1, 2, 3];

hilti::print(|xs|);
}

$ hiltic -j f.hlt
In file included from /private/var/folders/ht/vtb0rkbd4ws15vx02xdb84zw0000gn/T/hilti-rt-test-DxwpeY/foo.cc:6:
In file included from /usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/libhilti.h:10:
In file included from /usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/configuration.h:10:
In file included from /usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/util.h:19:
In file included from /usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/exception.h:12:
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/extension-points.h:28:12: error: call to 'to_string' is ambiguous
    return to_string(std::forward<T>(x), detail::adl::tag{});
           ^~~~~~~~~
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/extension-points.h:34:23: note: in instantiation of function template specialization 'hilti::rt::to_string<const unsigned long &>' requested here
    return hilti::rt::to_string(x);
                      ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/extension-points.h:46:20: note: in instantiation of function template specialization 'hilti::rt::detail::to_string_for_print<unsigned long>' requested here
    return detail::to_string_for_print(x);
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/hilti.h:28:24: note: in instantiation of function template specialization 'hilti::rt::to_string_for_print<unsigned long>' requested here
    cout << hilti::rt::to_string_for_print(t);
                       ^
/private/var/folders/ht/vtb0rkbd4ws15vx02xdb84zw0000gn/T/hilti-rt-test-DxwpeY/foo.cc:30:16: note: in instantiation of function template specialization 'hilti::rt::print<unsigned long>' requested here
    hilti::rt::print(foo::__globals()->xs.size(), hilti::rt::Bool(true));
               ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/bool.h:23:20: note: candidate function
inline std::string to_string(bool x, adl::tag /*unused*/) { return (x ? "True" : "False"); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:55:20: note: candidate function
inline std::string to_string(uint64_t x, adl::tag /*unused*/) { return fmt("%" PRIu64, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:57:20: note: candidate function
inline std::string to_string(int64_t x, adl::tag /*unused*/) { return fmt("%" PRId64, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:59:20: note: candidate function
inline std::string to_string(uint32_t x, adl::tag /*unused*/) { return fmt("%" PRIu32, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:61:20: note: candidate function
inline std::string to_string(int32_t x, adl::tag /*unused*/) { return fmt("%" PRId32, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:63:20: note: candidate function
inline std::string to_string(uint16_t x, adl::tag /*unused*/) { return fmt("%" PRIu16, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:65:20: note: candidate function
inline std::string to_string(int16_t x, adl::tag /*unused*/) { return fmt("%" PRId16, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:67:20: note: candidate function
inline std::string to_string(uint8_t x, adl::tag /*unused*/) { return fmt("%" PRIu8, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/integer.h:69:20: note: candidate function
inline std::string to_string(int8_t x, adl::tag /*unused*/) { return fmt("%" PRId8, x); }
                   ^
/usr/local/Cellar/spicy/HEAD-c92a84d/include/hilti/rt/types/real.h:28:13: note: candidate function
std::string to_string(double x, tag /*unused*/);
            ^
1 error generated.
[error] JIT compilation failed
[error] (no error output)
```

I'll try to make this operator return a safe integer instead which we should handle well.

This issue seems to be similar across `Map`, `Set`, and `Vector`.

--- a/hilti/runtime/include/types/map.h
+++ b/hilti/runtime/include/types/map.h
@@ -22,6 +22,7 @@
 
 #include <hilti/rt/extension-points.h>
 #include <hilti/rt/iterator.h>
+#include <hilti/rt/safe-int.h>
 #include <hilti/rt/util.h>
 
 namespace hilti::rt {
@@ -176,7 +177,7 @@ public:
 
     using key_type = typename M::key_type;
     using value_type = typename M::value_type;
-    using size_type = uint64_t;
+    using size_type = integer::safe<uint64_t>;
 
     using iterator = typename map::Iterator<K, V>;
     using const_iterator = typename map::ConstIterator<K, V>;

--- a/hilti/runtime/include/types/set.h
+++ b/hilti/runtime/include/types/set.h
@@ -20,6 +20,7 @@
 
 #include <hilti/rt/extension-points.h>
 #include <hilti/rt/iterator.h>
+#include <hilti/rt/safe-int.h>
 #include <hilti/rt/types/set_fwd.h>
 #include <hilti/rt/types/vector_fwd.h>
 #include <hilti/rt/util.h>
@@ -120,7 +121,7 @@ public:
     using key_type = T;
     using value_type = T;
 
-    using size_type = uint64_t;
+    using size_type = integer::safe<uint64_t>;
 
     Set() = default;
     Set(const Set&) = default;

--- a/hilti/runtime/include/types/vector.h
+++ b/hilti/runtime/include/types/vector.h
@@ -31,6 +31,7 @@
 #include <hilti/rt/extension-points.h>
 #include <hilti/rt/fmt.h>
 #include <hilti/rt/iterator.h>
+#include <hilti/rt/safe-int.h>
 #include <hilti/rt/types/vector_fwd.h>
 #include <hilti/rt/util.h>
 
@@ -253,7 +254,7 @@ public:
 
     using V = std::vector<T, Allocator>;
 
-    using size_type = uint64_t;
+    using size_type = integer::safe<uint64_t>;
     using reference = T&;
     using const_reference = const T&;
     using iterator = vector::Iterator<T, Allocator>;
@@ -465,7 +466,7 @@ public:
     auto cbegin() const { return const_iterator(0u, _control); }
     auto cend() const { return const_iterator(size(), _control); }
 
-    size_t size() const { return V::size(); }
+    size_type size() const { return V::size(); }
 
     // Methods of `std::vector`.
     using typename V::value_type;

--- a/hilti/runtime/src/tests/map.cc
+++ b/hilti/runtime/src/tests/map.cc
@@ -41,7 +41,7 @@ TEST_CASE("subscript") {
         CHECK_THROWS_WITH_AS(m[99], "key is unset", const IndexError&);
         // Proxy objects only invalidate iterators if an element was actually inserted.
         m = Map<int, int>{{1, 11}};
-        REQUIRE(m.size());
+        REQUIRE_EQ(m.size(), 1u);
         auto begin = m.begin();
         REQUIRE_EQ(begin->first, 1);
         REQUIRE_EQ(begin->second, 11);
@@ -148,7 +148,7 @@ TEST_CASE("Iterator") {
 TEST_CASE("index_assign") {
     // Modifying an existing element does not invalidate iterators.
     auto m = Map<int, int>{{1, 11}};
-    REQUIRE(m.size());
+    REQUIRE_EQ(m.size(), 1u);
     auto begin = m.begin();
     REQUIRE_EQ(begin->first, 1);
     REQUIRE_EQ(begin->second, 11);

--- a/hilti/runtime/src/tests/set.cc
+++ b/hilti/runtime/src/tests/set.cc
@@ -62,7 +62,7 @@ TEST_CASE("erase") {
     REQUIRE_EQ(*it1, 1);
     REQUIRE_EQ(*it2, 2);
 
-    REQUIRE(s.erase(1));
+    REQUIRE_EQ(s.erase(1), 1u);
 
     // In contrast to a `std::set`, removing elements from a `Set` invalidates
     // _all iterators_, not just iterators to the removed element.

